File: chapter5/auth_backend/src/jwt_authorizer.ts
import _ from 'lodash';
import { decode, verify } from 'jsonwebtoken';
import { fetch } from 'fetch-h2';
import jwks from 'jwks-rsa';
import throwError from './throwError';
import throwException from './throwException';
import AbstractAuthorizer from './AbstractAuthorizer';

class JwtAuthorizer extends AbstractAuthorizer {
  private readonly oidcConfigUrl: string;
  private readonly rolesClaimPath: string;
  private readonly getUsersUrl: string;
  private jwksClient: any;

  constructor() {
    super();
    // URL for OpenId Connect configuration endpoint in the IAM system
    this.oidcConfigUrl =
      process.env.OIDC_CONFIG_URL ??
      throwException('OIDC_CONFIG_URL is not defined');

    // With Keycloak you can use e.g., realm_access.roles
    this.rolesClaimPath = process.env.JWT_ROLES_CLAIM_PATH ??
      throwException('JWT_ROLES_CLAIM_PATH is not defined');

    // This is the URL where you can fetch the user id for a
    // specific 'sub' claim value in the access token
    // For example: http://localhost:8082/user-service/users
    this.getUsersUrl =
      process.env.GET_USERS_URL ??
      throwException('GET_USERS_URL is not defined');
  }

  async authorize(authHeader: string | undefined): Promise<void> {
    await this.tryGetClaims(authHeader);
  }

  // Authorize a user for him/herself
  async authorizeForSelf(
    userId: number,
    authHeader: string | undefined
  ): Promise<void> {
    const userIdInJwt = await this.getUserId(authHeader);

    if (userId !== userIdInJwt) {
      throwError(JwtAuthorizer.UnauthorizedError);
    }
  }

  async authorizeIfUserHasOneOfRoles(
    allowedRoles: string[],
    authHeader: string | undefined
  ): Promise<void> {
    const claims = await this.tryGetClaims(authHeader);
    const roles = _.get(claims, this.rolesClaimPath, []);

    const isAuthorized = allowedRoles.some((allowedRole) =>
      roles.includes(allowedRole));

    if (!isAuthorized) {
      throwError(JwtAuthorizer.UnauthorizedError);
    }
  }

  async getUserId(authHeader: string | undefined): Promise<number> {
    const claims = await this.tryGetClaims(authHeader);
    const getUsersUrl = this.getUsersUrl + `?sub=${claims?.sub}&fields=id`;
    let users;

    try {
      const usersResponse = await fetch(getUsersUrl);
      users = await usersResponse.json();
    } catch (error) {
      // Log error details
      throw new JwtAuthorizer.IamError();
    }

    return users?.[0]?.id ?? throwError(JwtAuthorizer.UnauthorizedError);
  }

  private async tryGetClaims(authHeader: string | undefined): Promise<any> {
    if (!authHeader) {
      throwError(JwtAuthorizer.UnauthenticatedError);
    }

    let oidcConfig;

    try {
      const oidcConfigResponse = await fetch(this.oidcConfigUrl);
      oidcConfig = await oidcConfigResponse.json();
    } catch (error) {
      // Log error details
      throwError(JwtAuthorizer.IamError);
    }

    if (!this.jwksClient) {
      this.jwksClient = jwks({ jwksUri: oidcConfig?.jwks_uri });
    }

    const jwt = authHeader?.split('Bearer ').pop() ?? '';

    try {
      const decodedJwt = decode(jwt, { complete: true });
      const kid = decodedJwt?.header?.kid;
      const signingKey = await this.jwksClient.getSigningKey(kid);
      return verify(jwt, signingKey.getPublicKey());
    } catch (error) {
      throwError(JwtAuthorizer.UnauthorizedError);
    }
  }
}

export const authorizer = new JwtAuthorizer();

File: chapter2/orderservice/src/main/java/com/example/orderservice/common/authorizer/JwtAuthorizer.java
package com.example.orderservice.common.authorizer;

import com.auth0.jwk.JwkException;
import com.auth0.jwk.JwkProviderBuilder;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.exceptions.JWTVerificationException;
import com.auth0.jwt.interfaces.Claim;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

import java.security.interfaces.RSAPublicKey;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.TimeUnit;

// @Service
public class JwtAuthorizer implements Authorizer {
  private final String issuerUrl;
  private final String jwksUrl;
  private final String getUsersUrl;

  public JwtAuthorizer() {
    // https:{your-IAM-service-domain-name}
    this.issuerUrl = System.getenv("ISSUER_URL");
    // https:{your-IAM-service-domain-name}/.well-known/jwks.json
    this.jwksUrl = System.getenv("JWKS_URL");
    this.getUsersUrl = System.getenv("GET_USERS_URL");
  }

  @Override
  public void authorize(final Optional<String> authHeader) {
    this.tryGetClaims(authHeader);
  }

  @Override
  public void authorizeForSelf(final Optional<String> maybeUserId, final Optional<String> maybeAuthHeader) {
    final var userIdInJwt = this.getUserId(maybeAuthHeader);

    if (maybeUserId.isEmpty() || !maybeUserId.get().equals(userIdInJwt)) {
      throw new UnauthorizedError();
    }
  }

  @Override
  public void authorizeIfUserHasOneOfRoles(final List<String> allowedRoles, final Optional<String> maybeAuthHeader) {
    final var claims = this.tryGetClaims(maybeAuthHeader);
    final var roles = List.of(claims.get("roles").asArray(String.class));
    final var isAuthorized = allowedRoles.stream().anyMatch(roles::contains);

    if (!isAuthorized) {
      throw new UnauthorizedError();
    }
  }

  @Override
  public String getUserId(final Optional<String> maybeAuthHeader) {
    final var claims = this.tryGetClaims(maybeAuthHeader);
    final var getUserUrl = getUsersUrl + "?sub=" + claims.get("sub").asString() + "&fields=id";

    try {
      final var users = new RestTemplate().getForObject(getUserUrl, User[].class);

      if (users != null && users.length == 1 && users[0].getId() != null) {
        return users[0].getId();
      } else {
        throw new UnauthorizedError();
      }
    } catch (final RestClientException error) {
      // Log error details
      throw new IamError();
    }
  }

  private Map<String, Claim> tryGetClaims(final Optional<String> maybeAuthHeader) {
    final var maybeClaims = getMaybeJwt(maybeAuthHeader).map(jwt -> {
      try {
        final var jwkProvider = new JwkProviderBuilder(jwksUrl).cached(10, 48, TimeUnit.HOURS).build();
        final var decodedJwt = JWT.decode(jwt);
        final var jwk = jwkProvider.get(decodedJwt.getKeyId());
        final var algorithm = Algorithm.RSA256((RSAPublicKey) jwk.getPublicKey(), null);
        final var jwtVerifier = JWT.require(algorithm).withIssuer(issuerUrl).build();
        final var verifiedJWT = jwtVerifier.verify(jwt);
        return verifiedJWT.getClaims();
      } catch (final JwkException error) {
        throw new IamError();
      } catch (final JWTVerificationException error) {
        throw new UnauthorizedError();
      }
    });

    return maybeClaims.orElse(new HashMap<>());
  }

  private Optional<String> getMaybeJwt(final Optional<String> maybeAuthHeader) {
    return maybeAuthHeader.map(authHeader -> {
      final var authHeadersParts = authHeader.split("Bearer ");
      return authHeadersParts.length == 2 ? authHeadersParts[1] : null;
    });
  }
}
